patchers/value/value.go

```diff
@@ -0,0 +1,260 @@
+	Uint64Value() uint64
+}
+
+type Float32Valuer interface {
+	Float32Value() float32
+}
+
+type Float64Valuer interface {
+	Float64Value() float64
+}
+
+type StringValuer interface {
+	StringValue() string
+}
+
+type TimeValuer interface {
+	TimeValue() time.Time
+}
+
+type DurationValuer interface {
+	DurationValue() time.Duration
+}
+
+type ArrayValuer interface {
+	ArrayValue() []any
+}
+
+type MapValuer interface {
+	MapValue() map[string]any
+}
+
+var supportedInterfaces = []reflect.Type{
+	reflect.TypeOf((*ExprValuer)(nil)).Elem(),
+	reflect.TypeOf((*BoolValuer)(nil)).Elem(),
+	reflect.TypeOf((*IntValuer)(nil)).Elem(),
+	reflect.TypeOf((*Int8Valuer)(nil)).Elem(),
+	reflect.TypeOf((*Int16Valuer)(nil)).Elem(),
+	reflect.TypeOf((*Int32Valuer)(nil)).Elem(),
+	reflect.TypeOf((*Int64Valuer)(nil)).Elem(),
+	reflect.TypeOf((*UintValuer)(nil)).Elem(),
+	reflect.TypeOf((*Uint8Valuer)(nil)).Elem(),
+	reflect.TypeOf((*Uint16Valuer)(nil)).Elem(),
+	reflect.TypeOf((*Uint32Valuer)(nil)).Elem(),
+	reflect.TypeOf((*Uint64Valuer)(nil)).Elem(),
+	reflect.TypeOf((*Float32Valuer)(nil)).Elem(),
+	reflect.TypeOf((*Float64Valuer)(nil)).Elem(),
+	reflect.TypeOf((*StringValuer)(nil)).Elem(),
+	reflect.TypeOf((*TimeValuer)(nil)).Elem(),
+	reflect.TypeOf((*DurationValuer)(nil)).Elem(),
+	reflect.TypeOf((*ArrayValuer)(nil)).Elem(),
+	reflect.TypeOf((*MapValuer)(nil)).Elem(),
+}
+
+type patcher struct{}
+
+func (patcher) Visit(node *ast.Node) {
+	id, ok := (*node).(*ast.IdentifierNode)
+	if !ok {
+		return
+	}
+
+	callnode := &ast.CallNode{
+		Callee:    &ast.IdentifierNode{Value: "getExprValue"},
+		Arguments: []ast.Node{id},
+	}
+
+	nodeType := id.Type()
+
+	for _, t := range supportedInterfaces {
+		if nodeType.Implements(t) {
+			ast.Patch(node, callnode)
+		}
+	}
+}
+
+func (patcher) ApplyOptions(c *conf.Config) {
+	getExprValueFunc(c)
+}
+
+func getExprValue(params ...any) (any, error) {
+
+	switch v := params[0].(type) {
+	case ExprValuer:
+		return v.ExprValue(), nil
+	case BoolValuer:
+		return v.BoolValue(), nil
+	case IntValuer:
+		return v.IntValue(), nil
+	case Int8Valuer:
+		return v.Int8Value(), nil
+	case Int16Valuer:
+		return v.Int16Value(), nil
+	case Int32Valuer:
+		return v.Int32Value(), nil
+	case Int64Valuer:
+		return v.Int64Value(), nil
+	case UintValuer:
+		return v.UintValue(), nil
+	case Uint8Valuer:
+		return v.Uint8Value(), nil
+	case Uint16Valuer:
+		return v.Uint16Value(), nil
+	case Uint32Valuer:
+		return v.Uint32Value(), nil
+	case Uint64Valuer:
+		return v.Uint64Value(), nil
+	case Float32Valuer:
+		return v.Float32Value(), nil
+	case Float64Valuer:
+		return v.Float64Value(), nil
+	case StringValuer:
+		return v.StringValue(), nil
+	case TimeValuer:
+		return v.TimeValue(), nil
+	case DurationValuer:
+		return v.DurationValue(), nil
+	case ArrayValuer:
+		return v.ArrayValue(), nil
+	case MapValuer:
+		return v.MapValue(), nil
+	}
+
+	return params[0], nil
+}
+
+var getExprValueFunc = expr.Function("getExprValue", getExprValue,
```

Comment:
Also let's rename `getExprValue` to something what will definitely not collide with user defined functions.

Like, `$patcher_value_getter`.

Reply:
c9797215aab8a4f000e1eeb1c00edf660c2b8592